iocore/utils/SharedExtendible.h

```diff
@@ -0,0 +1,447 @@
+/** @file
+
+  SharedAccess
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  @section details Details
+
+////////////////////////////////////////////
+  Implements SharedExtendible<Derived_t>:
+  * SharedExtendible<Derived_t>::Schema
+  * SharedExtendible<Derived_t>::FieldSchema
+  * SharedExtendible<Derived_t>::FieldId<Access_t,Field_t>
+ */
+
+#pragma once
+#include "stdint.h"
+#include <cstddef>
+#include <atomic>
+#include <cstring>
+#include <unordered_map>
+#include <type_traits>
+
+#include "ts/ink_memory.h"
+#include "ts/ink_assert.h"
+#include "SharedAccess.h"
+
+#ifndef ROUNDUP
+#define ROUNDUP(x, y) ((((x) + ((y)-1)) / (y)) * (y))
+#endif
+
+using SharedOffset_t = uint16_t;
+
+// used for C API
+typedef const void *FieldId_c;
+
+/// all types must allow unblocking MT read access
+enum SharedAccessEnum { ATOMIC, BIT, CONST, COPYSWAP, C_API, NUM_ACCESS_TYPES };
+
+/**
+ * @brief Allows code (and Plugins) to declare member variables during system init.
+ *
+ * The size of this structure is actually zero, so it will not change the size of your derived class.
+ * But new and delete are overriden to use allocate enough bytes of the derived type + added fields.
+ * All bool's are packed to save space using the *Bit methods.
+ * This API is focused on thread safe data types that allow minimally blocked reading.
+ * This is templated so static variables are instanced per Derived type. B/c we need to have different
+ * size field sets.
+ *
+ * @tparam Derived_t - the class that you want to extend at runtime.
+ *
+ * @see test_SharedExtendible.cc for examples
+ *
+ */
+template <typename Derived_t> struct SharedExtendible {
+  using SharedExtendible_t = SharedExtendible<Derived_t>;
```

Comment:
Isn't `SharedExtendible` available for this purpose inside the class?

Reply:
gone.